routing/missioncontrol.go

```diff
@@ -260,11 +261,58 @@ func (m *MissionControl) setLastPairResult(fromNode,
 		m.lastPairResult[fromNode] = nodePairs
 	}
 
-	nodePairs[toNode] = TimedPairResult{
-		Timestamp: timestamp,
-		Amt:       result.amt,
-		Success:   result.success,
+	current := nodePairs[toNode]
+
+	// Apply the new result to the existing data for this pair. If there is
+	// no existing data, apply it to the default values for TimedPairResult.
+	if result.success {
+		current.SuccessTime = timestamp
+		successAmt := result.amt
+
+		// Only update the success amount if this amount is higher. This
+		// prevents the success range from shrinking when there is no
+		// reason to do so. For example: small amount probes shouldn't
+		// affect a previous success for a much larger amount.
+		if successAmt > current.SuccessAmt {
+			current.SuccessAmt = successAmt
+		}
+
+		// If the success amount goes into the failure range, move the
+		// failure range up. Future attempts up to the success amount
+		// are likely to succeed. We don't want to clear the failure
+		// completely, because we haven't learnt much for amounts above
+		// the current success amount.
+		if !current.FailTime.IsZero() && successAmt >= current.FailAmt {
+			current.FailAmt = successAmt + 1
+		}
+	} else {
+		// For failures we always want to update both the amount and the
```

Comment:
I guess we have something similar here, with a failure to send 100k sats overwrites the information we had about failure to send 1 sat earlier. But this information is most likely hard to keep without actually storing all previous results (which might not be such a bad idea to do eventually?

Reply:
It may overwrite it, but it is unlikely to happen soon after the 1 sat failure. Because then the success probability estimate for that edge is still low.

I hope we will never have to store the full list of results and come up with some complicated formula to translate all of that into a probability. If we are lucky, the network quality improves enough to actually make this all less fine-grained.